File: SQL/Select/SQLConditions.vb

' ___________________________________________________
'
'  © Hi-Integrity Systems 2010. All rights reserved.
'  www.hisystems.com.au - Toby Wicks
' ___________________________________________________
'

Option Strict On
Option Explicit On 

Namespace SQL

    Public Class SQLConditions
        Implements IEnumerable

        Private pobjSQLConditions As New ArrayList
        Private pobjLogicalOperators As New Collections.Generic.List(Of LogicalOperator)

        Public Sub New()

        End Sub

        Public Function Add( _
            ByVal strFieldName As String, _
            ByVal eCompare As ComparisonOperator, _
            ByVal objValue As Object) As SQLCondition

            Return Add(strFieldName, eCompare, objValue, Nothing)

        End Function

        Public Function Add( _
            ByVal strFieldName As String, _
            ByVal eCompare As ComparisonOperator, _
            ByVal objValue As Object, _
            ByVal objTable As SQLSelectTable) As SQLCondition

            If strFieldName = String.Empty Then
                Throw New ArgumentNullException("Fieldname is null")
            End If

            EnsurePreviousLogicalOperatorExists()

            Add = New SQLCondition

            Add.Table = objTable
            Add.FieldName = strFieldName
            Add.Compare = eCompare
            Add.Value = objValue

            pobjSQLConditions.Add(Add)

        End Function

        Public Sub Add( _
            ByVal objCondition As SQLCondition)

            Me.Add(objCondition.FieldName, objCondition.Compare, objCondition.Value, objCondition.Table)

        End Sub

        Public Sub Add( _
            ByVal objConditions As SQLConditions)

            If objConditions.IsEmpty Then
                Throw New ArgumentException("SQLConditions does not contain any conditions.")
            End If

            EnsurePreviousLogicalOperatorExists()
            pobjSQLConditions.Add(objConditions)

        End Sub

        'Public Function Add( _
        '    ByVal strFieldName As String, _
        '    ByVal eCompare As ComparisonOperator, _
        '    ByVal objRightExpression As SQLExpression) As SQLConditionExpression

        '    Return Me.Add(New SQLFieldExpression(strFieldName), eCompare, objRightExpression)

        'End Function

        Public Function Add( _
            ByVal objLeftExpression As SQLExpression, _
            ByVal eCompare As ComparisonOperator, _
            ByVal objRightExpression As SQLExpression) As SQLConditionExpression

            Dim objCondition As New SQLConditionExpression(objLeftExpression, eCompare, objRightExpression)

            Me.Add(objCondition)

            Return objCondition

        End Function

        Public Sub Add(ByVal objCondition As SQLConditionExpression)

            If objCondition Is Nothing Then
                Throw New ArgumentNullException
            End If

            EnsurePreviousLogicalOperatorExists()
            pobjSQLConditions.Add(objCondition)

        End Sub

        Public Function Add() As SQLCondition

            Return Add("", ComparisonOperator.EqualTo, Nothing, Nothing)

        End Function

        Public Function AddInSelect() As SQLConditionInSelect

            Return AddInSelect(String.Empty, Nothing, Nothing)

        End Function

        Public Function AddInSelect( _
            ByVal strFieldName As String, _
            ByVal objSelect As SQLSelect) As SQLConditionInSelect

            Return AddInSelect(strFieldName, objSelect, Nothing)

        End Function

        Public Function AddInSelect( _
            ByVal strFieldName As String, _
            ByVal objSelect As SQLSelect, _
            ByVal objTable As SQLSelectTable) As SQLConditionInSelect

            EnsurePreviousLogicalOperatorExists()

            AddInSelect = New SQLConditionInSelect
            AddInSelect.Table = objTable
            AddInSelect.FieldName = strFieldName
            AddInSelect.Select = objSelect

            pobjSQLConditions.Add(AddInSelect)

        End Function

        Public Function AddNotInSelect() As SQLConditionInSelect

            Return AddNotInSelect(String.Empty, Nothing, Nothing)

        End Function

        Public Function AddNotInSelect( _
            ByVal strFieldName As String, _
            ByVal objSelect As SQLSelect) As SQLConditionInSelect

            Return AddNotInSelect(strFieldName, objSelect, Nothing)

        End Function

        Public Function AddNotInSelect( _
            ByVal strFieldName As String, _
            ByVal objSelect As SQLSelect, _
            ByVal objTable As SQLSelectTable) As SQLConditionInSelect

            AddNotInSelect = AddInSelect(strFieldName, objSelect, objTable)
            AddNotInSelect.NotInSelect = True

        End Function

        Public Function AddSelect() As SQLConditionSelect

            Return AddSelect(Nothing, ComparisonOperator.EqualTo, Nothing)

        End Function

        Public Function AddSelect( _
            ByVal objSelect As SQLSelect, _
            ByVal eCompare As ComparisonOperator, _
            ByVal objValue As Object) As SQLConditionSelect

            EnsurePreviousLogicalOperatorExists()

            AddSelect = New SQLConditionSelect
            AddSelect.Select = objSelect
            AddSelect.Compare = eCompare
            AddSelect.Value = objValue

            pobjSQLConditions.Add(AddSelect)

        End Function

        Public Function AddFieldCompare() As SQLConditionFieldCompare

            Return AddFieldCompare(Nothing, "", ComparisonOperator.EqualTo, Nothing, "")

        End Function

        Public Function AddFieldCompare( _
            ByVal strFieldName1 As String, _
            ByVal eCompare As ComparisonOperator, _
            ByVal objTable2 As SQLSelectTable, _
            ByVal strFieldName2 As String) As SQLConditionFieldCompare

            Return AddFieldCompare(Nothing, strFieldName1, eCompare, objTable2, strFieldName2)

        End Function

        Public Function AddFieldCompare( _
            ByVal objTable1 As SQLSelectTable, _
            ByVal strFieldName1 As String, _
            ByVal eCompare As ComparisonOperator, _
            ByVal objTable2 As SQLSelectTable, _
            ByVal strFieldName2 As String) As SQLConditionFieldCompare

            EnsurePreviousLogicalOperatorExists()

            AddFieldCompare = New SQLConditionFieldCompare

            With AddFieldCompare
                .Table1 = objTable1
                .FieldName1 = strFieldName1
                .Compare = eCompare
                .Table2 = objTable2
                .FieldName2 = strFieldName2
            End With

            pobjSQLConditions.Add(AddFieldCompare)

        End Function

        Public ReadOnly Property IsEmpty() As Boolean
            Get

                Return pobjSQLConditions.Count = 0

            End Get
        End Property

        Private Sub EnsurePreviousLogicalOperatorExists()

            'Add the AND operator if an operator hasn't been called after the previous Add call
            If pobjLogicalOperators.Count() < pobjSQLConditions.Count() Then
                Me.AddLogicalOperator(LogicalOperator.And)
            End If

        End Sub

        Public Sub AddLogicalOperator( _
            ByVal eLogicalOperator As LogicalOperator)

            If pobjLogicalOperators.Count() + 1 > pobjSQLConditions.Count() Then
                Throw New Exceptions.DatabaseObjectsException("First call the Add function - this function has been called without a prior call to Add")
            End If

            pobjLogicalOperators.Add(eLogicalOperator)

        End Sub

        Public Sub Delete(ByRef objCondition As SQLCondition)

            Dim intConditionIndex As Integer = pobjSQLConditions.IndexOf(objCondition)

            If Not pobjSQLConditions.Contains(objCondition) Then
                Throw New IndexOutOfRangeException
            End If

            If intConditionIndex > 0 Then
                pobjLogicalOperators.Remove(pobjLogicalOperators(intConditionIndex - 1))
            ElseIf intConditionIndex = 0 Then
                pobjLogicalOperators.Remove(pobjLogicalOperators(0))
            End If

            pobjSQLConditions.Remove(objCondition)
            objCondition = Nothing

        End Sub

        Friend ReadOnly Property SQL(ByVal eConnectionType As Database.ConnectionType) As String
            Get

                Dim strSQL As String = String.Empty

                For intIndex As Integer = 0 To pobjSQLConditions.Count() - 1
                    If intIndex > 0 Then
                        strSQL &= " " & SQLConvertLogicalOperator(pobjLogicalOperators.Item(intIndex - 1)) & " "
                    End If

                    If TypeOf pobjSQLConditions(intIndex) Is SQLCondition Then
                        strSQL &= DirectCast(pobjSQLConditions(intIndex), SQLCondition).SQL(eConnectionType)
                    ElseIf TypeOf pobjSQLConditions(intIndex) Is SQLConditions Then
                        strSQL &= "(" & DirectCast(pobjSQLConditions(intIndex), SQLConditions).SQL(eConnectionType) & ")"
                    ElseIf TypeOf pobjSQLConditions(intIndex) Is SQLConditionInSelect Then
                        strSQL &= DirectCast(pobjSQLConditions(intIndex), SQLConditionInSelect).SQL(eConnectionType)
                    ElseIf TypeOf pobjSQLConditions(intIndex) Is SQLConditionSelect Then
                        strSQL &= DirectCast(pobjSQLConditions(intIndex), SQLConditionSelect).SQL(eConnectionType)
                    ElseIf TypeOf pobjSQLConditions(intIndex) Is SQLConditionFieldCompare Then
                        strSQL &= DirectCast(pobjSQLConditions(intIndex), SQLConditionFieldCompare).SQL(eConnectionType)
                    ElseIf TypeOf pobjSQLConditions(intIndex) Is SQLConditionExpression Then
                        strSQL &= DirectCast(pobjSQLConditions(intIndex), SQLConditionExpression).SQL(eConnectionType)
                    Else
                        Throw New NotImplementedException(pobjSQLConditions(intIndex).GetType.FullName)
                    End If
                Next

                Return strSQL

            End Get
        End Property

        Private Function GetEnumerator() As System.Collections.IEnumerator Implements System.Collections.IEnumerable.GetEnumerator

            Return pobjSQLConditions.GetEnumerator

        End Function

    End Class

End Namespace

File: SQL/Select/SQLConditions.cs
// _________________________________________________________________________
//
//  © Hi-Integrity Systems 2010. All rights reserved.
//  www.hisystems.com.au - Toby Wicks
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//	    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// _________________________________________________________________________
//

using System.Collections;
using System;
using System.Data;
using System.Collections.Generic;
using DatabaseObjects.SQL.Serializers;

namespace DatabaseObjects.SQL
{
	public class SQLConditions : SQLExpression, IEnumerable
	{
		private ArrayList pobjSQLConditions = new ArrayList();
		private List<LogicalOperator> pobjLogicalOperators = new List<LogicalOperator>();
			
		public SQLConditions()
		{
		}
			
		public SQLCondition Add(string strFieldName, ComparisonOperator eCompare, object objValue)
		{
			return Add(strFieldName, eCompare, objValue, null);
		}
			
		public SQLCondition Add(string strFieldName, ComparisonOperator eCompare, object objValue, SQLSelectTable objTable)
		{
			if (String.IsNullOrEmpty(strFieldName))
				throw new ArgumentNullException("Fieldname is null");
				
			EnsurePreviousLogicalOperatorExists();
				
			var condition = new SQLCondition();
				
			condition.Table = objTable;
			condition.FieldName = strFieldName;
			condition.Compare = eCompare;
			condition.Value = objValue;
				
			pobjSQLConditions.Add(condition);
				
			return condition;
		}
			
		public void Add(SQLCondition objCondition)
		{
			this.Add(objCondition.FieldName, objCondition.Compare, objCondition.Value, objCondition.Table);
		}
			
		public void Add(SQLConditions objConditions)
		{
			if (objConditions.IsEmpty)
				throw new ArgumentException("SQLConditions does not contain any conditions.");
				
			EnsurePreviousLogicalOperatorExists();
			pobjSQLConditions.Add(objConditions);
		}
			
		public SQLConditionExpression Add(SQLExpression objLeftExpression, ComparisonOperator eCompare, SQLExpression objRightExpression)
		{
			SQLConditionExpression objCondition = new SQLConditionExpression(objLeftExpression, eCompare, objRightExpression);
				
			this.Add(objCondition);
				
			return objCondition;
		}
			
		public void Add(SQLConditionExpression objCondition)
		{
			if (objCondition == null)
				throw new ArgumentNullException();
				
			EnsurePreviousLogicalOperatorExists();
			pobjSQLConditions.Add(objCondition);
		}
			
		public SQLCondition Add()
		{
			return Add("", ComparisonOperator.EqualTo, null, null);
		}
			
		public SQLConditionInSelect AddInSelect()
		{
			return AddInSelect(string.Empty, null, null);
		}
			
		public SQLConditionInSelect AddInSelect(string strFieldName, SQLSelect objSelect)
		{
			return AddInSelect(strFieldName, objSelect, null);
		}
			
		public SQLConditionInSelect AddInSelect(string strFieldName, SQLSelect objSelect, SQLSelectTable objTable)
		{
			EnsurePreviousLogicalOperatorExists();
				
			var condition = new SQLConditionInSelect();
			condition.Table = objTable;
			condition.FieldName = strFieldName;
			condition.Select = objSelect;
				
			pobjSQLConditions.Add(condition);
				
			return condition;
		}
			
		public SQLConditionInSelect AddNotInSelect()
		{
			return AddNotInSelect(string.Empty, null, null);
		}
			
		public SQLConditionInSelect AddNotInSelect(string strFieldName, SQLSelect objSelect)
		{
			return AddNotInSelect(strFieldName, objSelect, null);
		}
			
		public SQLConditionInSelect AddNotInSelect(string strFieldName, SQLSelect objSelect, SQLSelectTable objTable)
		{
			var condition = AddInSelect(strFieldName, objSelect, objTable);
			condition.NotInSelect = true;
				
			return condition;
		}
			
		public SQLConditionSelect AddSelect()
		{
			return AddSelect(null, ComparisonOperator.EqualTo, null);
		}
			
		public SQLConditionSelect AddSelect(SQLSelect objSelect, ComparisonOperator eCompare, object objValue)
		{
			EnsurePreviousLogicalOperatorExists();
				
			var condition = new SQLConditionSelect();
			condition.Select = objSelect;
			condition.Compare = eCompare;
			condition.Value = objValue;
				
			pobjSQLConditions.Add(condition);
				
			return condition;
		}
			
		public SQLConditionFieldCompare AddFieldCompare()
		{
			return AddFieldCompare(null, "", ComparisonOperator.EqualTo, null, "");
		}
			
		public SQLConditionFieldCompare AddFieldCompare(string strFieldName1, ComparisonOperator eCompare, SQLSelectTable objTable2, string strFieldName2)
		{
			return AddFieldCompare(null, strFieldName1, eCompare, objTable2, strFieldName2);
		}
			
		public SQLConditionFieldCompare AddFieldCompare(SQLSelectTable objTable1, string strFieldName1, ComparisonOperator eCompare, SQLSelectTable objTable2, string strFieldName2)
		{
			EnsurePreviousLogicalOperatorExists();
				
			var condition = new SQLConditionFieldCompare();
            condition.Table1 = objTable1;
            condition.FieldName1 = strFieldName1;
            condition.Compare = eCompare;
            condition.Table2 = objTable2;
            condition.FieldName2 = strFieldName2;
				
			pobjSQLConditions.Add(condition);
				
			return condition;
		}
			
		public bool IsEmpty
		{
			get
			{
				return pobjSQLConditions.Count == 0;
			}
		}
			
		private void EnsurePreviousLogicalOperatorExists()
		{
			//Add the AND operator if an operator hasn't been called after the previous Add call
			if (pobjLogicalOperators.Count < pobjSQLConditions.Count)
				this.AddLogicalOperator(LogicalOperator.And);
		}
			
		public void AddLogicalOperator(LogicalOperator eLogicalOperator)
		{
			if (pobjLogicalOperators.Count + 1 > pobjSQLConditions.Count)
				throw new Exceptions.DatabaseObjectsException("First call the Add function - this function has been called without a prior call to Add");
				
			pobjLogicalOperators.Add(eLogicalOperator);
		}
			
		public void Delete(ref SQLCondition objCondition)
		{
			int intConditionIndex = pobjSQLConditions.IndexOf(objCondition);
				
			if (!pobjSQLConditions.Contains(objCondition))
				throw new IndexOutOfRangeException();
				
			if (intConditionIndex > 0)
				pobjLogicalOperators.Remove(pobjLogicalOperators[intConditionIndex - 1]);
			else if (intConditionIndex == 0)
				pobjLogicalOperators.Remove(pobjLogicalOperators[0]);
				
			pobjSQLConditions.Remove(objCondition);
			objCondition = null;
		}

		internal LogicalOperator[] LogicalOperators
		{
			get
			{
				return this.pobjLogicalOperators.ToArray();
			}
		}

		public System.Collections.IEnumerator GetEnumerator()
		{
			return pobjSQLConditions.GetEnumerator();
		}

		internal override string SQL(Serializer serializer)
		{
			return serializer.SerializeConditions(this);
		}
	}
}
